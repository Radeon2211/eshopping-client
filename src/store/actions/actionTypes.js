export const LOGIN_USER = 'LOGIN_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const SET_PLACED_ORDERS = 'SET_PLACED_ORDERS';
export const SET_SELL_HISTORY = 'SET_SELL_HISTORY';
export const SET_PROFILE = 'SET_PROFILE';
export const UPDATE_TRANSACTION = 'UPDATE_TRANSACTION';
export const SET_PRODUCTS = 'SET_PRODUCTS';
export const SET_PRODUCT_DETAILS = 'SET_PRODUCT_DETAILS';
export const DELETE_PRODUCT_DETAILS = 'DELETE_PRODUCT_DETAILS';
export const DELETE_PRODUCT_FROM_LIST = 'DELETE_PRODUCT_FROM_LIST';
export const CLEAR_PRODUCTS = 'CLEAR_PRODUCTS';
export const FORM_START = 'FORM_START';
export const FORM_SUCCESS = 'FORM_SUCCESS';
export const FORM_FAIL = 'FORM_FAIL';
export const DATA_START = 'DATA_START';
export const DATA_SUCCESS = 'DATA_SUCCESS';
export const DATA_FAIL = 'DATA_FAIL';
export const SET_MESSAGE = 'SET_MESSAGE';
export const DELETE_MESSAGE = 'DELETE_MESSAGE';
export const SET_MODAL_OPEN_STATE = 'SET_MODAL_OPEN_STATE';
export const SET_MAX_QUANTITY_PER_PAGE = 'SET_MAX_QUANTITY_PER_PAGE';
