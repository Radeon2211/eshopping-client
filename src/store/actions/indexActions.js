export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
