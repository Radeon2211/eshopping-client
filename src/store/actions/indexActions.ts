export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeEmail,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions/authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions/tradeActions';

export { setMessage, setModal, changeProductsPerPage } from './uiActions/uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions/productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions/orderActions';
