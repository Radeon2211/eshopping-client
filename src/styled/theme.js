const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
