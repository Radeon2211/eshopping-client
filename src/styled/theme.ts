const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#777',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#06AD96',
    greenDark: '#009680',
    red: '#F22613',
    redLight: '#F94D2A',
    redDark: '#985A5A',
  },
  shadows: {
    level1: '0 0 3px rgba(0, 0, 0, 0.03)',
    level2: '0 0 3px rgba(0, 0, 0, 0.3)',
    level3: '0 0 5px rgba(0, 0, 0, 0.3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Noto Sans, sans-serif',
    text: 'Nunito, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
    level5: 100000,
  },
  fontSizes: {
    level1: '1.3rem',
    level2: '1.4rem',
    level3: '1.5rem',
    level4: '1.7rem',
    level5: '2.2rem',
    level6: '2.6rem',
  },
  durations: {
    level1: 0.1,
    level2: 0.2,
    level3: 0.3,
  },
  lineHeights: {
    level1: '1',
    level2: '1.1',
    level3: '1.2',
    level4: '1.3',
    level5: '1.4',
    level6: '1.5',
  },
};

export type ThemeSpacing = keyof typeof theme.spacings;
export type ThemeZIndex = keyof typeof theme.zIndexes;
export type ThemeFontSize = keyof typeof theme.fontSizes;
export type ThemeDuration = keyof typeof theme.durations;
export type ThemeLineHeight = keyof typeof theme.lineHeights;

export default theme;
