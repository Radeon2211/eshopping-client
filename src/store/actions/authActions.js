import axios from '../../axios';
import * as actionTypes from './actionTypes';
import * as uiActions from './uiActions';
import { getErrorMessage } from '../../shared/utility';
import { DEFAULT_PATH } from '../../shared/constants';

export const setProfile = (profile) => ({
  type: actionTypes.SET_PROFILE,
  profile,
});

export const logout = () => ({
  type: actionTypes.LOGOUT_USER,
});

export const setDeliveryAddress = (deliveryAddress) => ({
  type: actionTypes.SET_DELIVERY_ADDRESS,
  deliveryAddress,
});

export const setOtherUser = (user) => ({
  type: actionTypes.SET_OTHER_USER,
  otherUser: user,
});

export const registerUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    const country = creds.country.value;
    const phone = `+${creds.phonePrefix.value} ${creds.phoneNumber}`;
    const contacts = [];
    if (!creds.hideEmail) contacts.push('email');
    if (!creds.hidePhone) contacts.push('phone');
    const correctCreds = {
      ...creds,
      country,
      phone,
      contacts,
    };
    delete correctCreds.hideEmail;
    delete correctCreds.hidePhone;
    delete correctCreds.phoneNumber;
    delete correctCreds.phonePrefix;
    try {
      const { data } = await axios.post('/users', { data: correctCreds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage('Your account has been created successfully!'));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const loginUser = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.post('/users/login', { data: creds });
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchProfile = () => {
  return async (dispatch) => {
    try {
      const { data } = await axios.get('/users/me');
      dispatch(setProfile(data.user));
      dispatch(uiActions.writeChangeCartInfo(data.isDifferent));
    } catch (error) {
      dispatch(setProfile(null));
    }
  };
};

export const logoutUser = () => {
  return async (dispatch) => {
    try {
      await axios.post('/users/logout');
      dispatch(setProfile(null));
    } catch (error) {
      dispatch(uiActions.setAndDeleteMessage('Unable to logout. Something went wrong'));
    }
  };
};

export const updateUser = (creds, message) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { data } = await axios.patch('/users/me', creds);
      dispatch(setProfile(data.user));
      dispatch(uiActions.setAndDeleteMessage(message));
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const deleteAccount = (creds, history) => {
  return async (dispatch, getState) => {
    dispatch(uiActions.formStart());
    try {
      const { username } = getState().auth.profile;
      await axios.delete('/users/me', { data: creds });
      dispatch(setProfile(null));
      dispatch(
        uiActions.setAndDeleteMessage(`Your account has been deleted. Goodbye ${username}!`),
      );
      dispatch(uiActions.formSuccess());
      history.replace(DEFAULT_PATH);
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const changeDeliveryAddress = (creds) => {
  return async (dispatch) => {
    dispatch(uiActions.formStart());
    try {
      const { onlyTheseOrders } = creds;
      // eslint-disable-next-line no-param-reassign
      delete creds.onlyTheseOrders;

      if (onlyTheseOrders) {
        dispatch(setDeliveryAddress(creds));
      } else {
        const { data } = await axios.patch('/users/me', creds);
        dispatch(setProfile(data.user));
        dispatch(uiActions.setAndDeleteMessage('Delivery address has been saved in your profile'));
      }
      dispatch(uiActions.formSuccess());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.formFail(errorMessage));
    }
  };
};

export const fetchOtherUser = (username) => {
  return async (dispatch) => {
    dispatch(uiActions.dataStart());
    try {
      const { data } = await axios.get(`/users/${username}`);
      dispatch(setOtherUser(data.profile));
      dispatch(uiActions.dataEnd());
    } catch (error) {
      const errorMessage = getErrorMessage(error);
      dispatch(uiActions.setAndDeleteMessage(errorMessage));
      dispatch(setOtherUser(null));
      dispatch(uiActions.dataEnd());
    }
  };
};
