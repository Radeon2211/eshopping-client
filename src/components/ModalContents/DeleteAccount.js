import React, { useCallback } from 'react';
import { Formik } from 'formik';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import * as actions from '../../store/actions/indexActions';
import Form from '../UI/Form/Form';
import Input from '../UI/Input/Input';
import { inputKinds } from '../../shared/constants';

const DeleteAccount = () => {
  const history = useHistory();

  const dispatch = useDispatch();
  const onDeleteAccount = useCallback(
    (creds, currentHistory) => dispatch(actions.deleteAccount(creds, currentHistory)),
    [dispatch],
  );

  return (
    <Formik
      initialValues={{
        currentPassword: '',
      }}
      onSubmit={(data) => {
        onDeleteAccount(data, history);
      }}
    >
      {({ dirty }) => (
        <Form
          btnText="delete"
          btnColor="red"
          headingText="Delete your account"
          isValid={dirty}
          cancellable
        >
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'password',
              name: 'currentPassword',
              id: 'currentPassword',
              placeholder: 'Your current password',
              autoComplete: 'off',
            }}
            label="Current password"
          />
        </Form>
      )}
    </Formik>
  );
};

export default DeleteAccount;
