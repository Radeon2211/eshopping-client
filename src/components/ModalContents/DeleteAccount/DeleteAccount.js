import React, { useCallback } from 'react';
import { Formik } from 'formik';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import * as actions from '../../../store/actions/indexActions';
import Form from '../../UI/Form/Form';
import Input from '../../UI/Input/Input';
import { inputKinds } from '../../../shared/constants';

export default function DeleteAccount() {
  const navigate = useNavigate();

  const dispatch = useDispatch();
  const onDeleteAccount = useCallback(
    (creds, navigateFn) => dispatch(actions.deleteAccount(creds, navigateFn)),
    [dispatch],
  );

  return (
    <Formik
      initialValues={{
        currentPassword: '',
      }}
      onSubmit={(data) => {
        onDeleteAccount(data, navigate);
      }}
    >
      {({ dirty }) => (
        <Form
          btnText="delete"
          btnColor="red"
          headingText="Delete your account"
          isValid={dirty}
          cancellable
        >
          <Input
            kind={inputKinds.INPUT}
            config={{
              type: 'password',
              name: 'currentPassword',
              id: 'currentPassword',
              placeholder: 'Your current password',
              autoComplete: 'off',
              autoFocus: true,
              'data-testid': 'DeleteAccount-current-password',
            }}
            label="Current password"
          />
        </Form>
      )}
    </Formik>
  );
}
