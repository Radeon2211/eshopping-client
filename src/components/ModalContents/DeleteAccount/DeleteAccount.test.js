import React from 'react';
import { render, cleanup, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import DeleteAccount from './DeleteAccount';
import theme from '../../../styled/theme';
import * as actions from '../../../store/actions/indexActions';

const mockStore = configureMockStore([thunk]);

const setUp = () => {
  const store = mockStore({
    ui: {
      isFormLoading: false,
      formError: '',
    },
  });
  store.dispatch = jest.fn();

  const history = {
    listen: jest.fn(),
    location: { pathname: '/my-account/data' },
  };

  return {
    ...render(
      <Router history={history}>
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <DeleteAccount />
          </ThemeProvider>
        </Provider>
      </Router>,
    ),
    store,
    history,
  };
};

jest.mock('../../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../../store/actions/indexActions.js'),
  deleteAccount: (credentials, currentHistory) => ({
    credentials,
    currentHistory,
  }),
}));

afterEach(cleanup);

describe('<DeleteAccount />', () => {
  describe('Check how renders', () => {
    it('should render everything correctly', () => {
      const { asFragment } = setUp();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Check form', () => {
    it('should have empty input and focus on it by default', () => {
      setUp();
      const currentPasswordInput = screen.getByTestId('DeleteAccount-current-password');
      expect(currentPasswordInput).toHaveFocus();
      expect(currentPasswordInput.value).toEqual('');
    });

    it('should call deleteAccount() with given password after input submit and button click', async () => {
      const { store, history, container } = setUp();

      const currentPasswordInput = screen.getByTestId('DeleteAccount-current-password');
      const currentPassword = 'password';

      await waitFor(() => {
        fireEvent.change(currentPasswordInput, { target: { value: currentPassword } });
      });
      expect(currentPasswordInput.value).toEqual(currentPassword);

      await waitFor(() => {
        fireEvent.click(container.querySelector('button[type="submit"]'));
      });
      expect(store.dispatch).toHaveBeenNthCalledWith(
        1,
        actions.deleteAccount({ currentPassword }, history),
      );

      await waitFor(() => {
        fireEvent.submit(currentPasswordInput);
      });
      expect(store.dispatch).toHaveBeenNthCalledWith(
        2,
        actions.deleteAccount({ currentPassword }, history),
      );

      expect(store.dispatch).toHaveBeenCalledTimes(2);
    });

    it('should NOT call deleteAccount() if input is empty', async () => {
      const { store, container } = setUp();
      await waitFor(() => {
        fireEvent.click(container.querySelector('button[type="submit"]'));
      });
      expect(store.dispatch).not.toHaveBeenCalled();
    });
  });
});
