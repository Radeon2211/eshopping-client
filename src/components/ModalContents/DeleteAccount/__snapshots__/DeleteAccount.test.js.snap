// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteAccount /> Check how renders should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hYjfqi"
  >
    <h3
      class="sc-iBPRYJ kqmSLn"
    >
      Delete your account
    </h3>
    <form
      action="#"
      class="formik-form"
    >
      <div
        class="sc-fubCfw drZTsR"
        type="password"
      >
        <label
          class="sc-jrAGrp bkXExJ"
          for="currentPassword"
        >
          Current password
        </label>
        <input
          autocomplete="off"
          class="sc-pFZIQ fjonmB"
          data-testid="DeleteAccount-current-password"
          id="currentPassword"
          name="currentPassword"
          placeholder="Your current password"
          type="password"
          value=""
        />
      </div>
      <div
        class="sc-gKsewC dLOeOj"
        direction="row"
        spacing="3"
        width="100%"
        wrap="nowrap"
      >
        <span
          class="sc-eCssSg gABWUT"
          color=""
          display="block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          <button
            class="sc-gsTCUz cpcmUx"
            color="blue"
            type="button"
          >
            cancel
          </button>
        </span>
        <button
          class="sc-gsTCUz KHXhw"
          color="red"
          data-testid="Form-submit-btn"
          disabled=""
          type="submit"
        >
          delete
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
